Review and fix RBAC black agent API integration test

|Component|
|---|
| API Integration Tests |

The RBAC black agent API integration test is failing in the test case "Try to reconnect all agents (Partially allowed, user agnostic)". In this test case, we expect to reconnect all agents. The test is failing because the agent with ID 005 is not in the response and it should be. This agent may not be in the response because it is not active (we don't know this as the code related to non-active agents is excluded) as it was reconnected in the test case before this one.

There is a delay to avoid this race condition, maybe the delay is not enough to avoid it.

Test failing in Wazuh `master` branch.

<details>

<summary>Error output</summary>

```
Format variables:
  protocol:s = 'https'
  host:s = 'localhost'
  port:d = '55000'
  test_login_token = 'eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjM3MTkwNjI4LCJleHAiOjE2MzcxOTE1MjgsInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbOTldLCJyYmFjX21vZGUiOiJibGFjayJ9.AWNfuDlgkG5MWw8s9wBgSdXNqELRG9bdZ6v0XbAP6Db-KZgfuCUa7Dkx687WKASinxgKftLcDQmhUTqRKjMacFC-AZhmBsJ46gYmln-q8Xgf57VRTQTDgWvlOF-wMM34sMJu4IwajdCGzDn8FZRHdEgtCFKP0pOIhyMhKil7wo2huJzw'
  reconnect_delay = '30'

Source test stage (line 1395):
  - name: Try to reconnect all agents (Partially allowed, user agnostic)
    request:
      verify: False
      url: "{protocol:s}://{host:s}:{port:d}/agents/reconnect"
      method: PUT
      headers:
        Authorization: "Bearer {test_login_token}"
    response:
      status_code: 200
      json:
        error: 0
        data:
          affected_items:
            - '003'
            - '005'
            - '007'
          failed_items: []
          total_affected_items: 3
          total_failed_items: 0
    delay_after: !float "{reconnect_delay}"


Formatted stage:
  delay_after: !float '30'
  name: Try to reconnect all agents (Partially allowed, user agnostic)
  request:
    headers:
      Authorization: Bearer eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjM3MTkwNjI4LCJleHAiOjE2MzcxOTE1MjgsInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbOTldLCJyYmFjX21vZGUiOiJibGFjayJ9.AWNfuDlgkG5MWw8s9wBgSdXNqELRG9bdZ6v0XbAP6Db-KZgfuCUa7Dkx687WKASinxgKftLcDQmhUTqRKjMacFC-AZhmBsJ46gYmln-q8Xgf57VRTQTDgWvlOF-wMM34sMJu4IwajdCGzDn8FZRHdEgtCFKP0pOIhyMhKil7wo2huJzw
    method: PUT
    url: 'https://localhost:55000/agents/reconnect'
    verify: false
  response:
    json:
      data:
        affected_items:
        - '003'
        - '005'
        - '007'
        failed_items: []
        total_affected_items: 3
        total_failed_items: 0
      error: 0
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Try to reconnect all agents (Partially allowed, user agnostic)' failed:
    - List item(s) not present in response: ['005', '007']
------------------------------ Captured log call -------------------------------
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '403'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
ERROR    tavern.response.base:base.py:41 List item(s) not present in response: ['005', '007']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'data': {'af...o all agents'} == {'data': {'af...}, 'error': 0}
  Omitting 1 identical items, use -vv to show
  Differing items:
  {'data': {'affected_items': ['003', '007'], 'failed_items': [], 'total_affected_items': 2, 'total_failed_items': 0}} != {'data': {'affected_items': ['003', '005', '007'], 'failed_items': [], 'total_affected_items': 3, 'total_failed_items': 0}}
  Left contains 1 more item:
  {'message': 'Force reconnect command was sent to all agents'}
  Full diff:
    {...
  
  ...Full output truncated (13 lines hidden), use '-vv' to show

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'affected_it...led_items': 0} == {'affected_it...led_items': 0}
  Omitting 2 identical items, use -vv to show
  Differing items:
  {'affected_items': ['003', '007']} != {'affected_items': ['003', '005', '007']}
  {'total_affected_items': 2} != {'total_affected_items': 3}
  Full diff:
    {
  -  'affected_items': ['003', '005', '007'],...
  
  ...Full output truncated (10 lines hidden), use '-vv' to show

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert ['003', '007'] == ['003', '005', '007']
  At index 1 diff: '007' != '005'
  Right contains one more item: '007'
  Full diff:
  - ['003', '005', '007']
  ?         -------
  + ['003', '007']

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/response/base.py", line 88, in recurse_check_key_match
    check_keys_match_recursive(expected_block, block, [], strict)
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 462, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 462, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 512, in check_keys_match_recursive
    raise exceptions.KeyMismatchError(msg) from e
tavern.util.exceptions.KeyMismatchError: List item(s) not present in response: ['005', '007']
```

</details>

We should review and fix this API integration test.

Also, review this possible race condition in the non-RBAC API integration test and RBAC white.

## Checks

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)
<!-- If changes are made to any of the following components, uncomment the corresponding line 
- [ ] **Integration tests** without failures for API configuration (`/wazuh-qa/tests/integration/test_api/test_config/`)
- [ ] **System tests** for agent enrollment process (`/wazuh-qa/tests/system/test_cluster/test_agent_enrollment`)
- [ ] **System tests** for agent info sync process in cluster (`/wazuh-qa/tests/system/test_cluster/test_agent_info_sync`)
- [ ] **System tests** for agent key polling (`/wazuh-qa/tests/system/test_cluster/test_agent_key_polling`)
- [ ] **System tests** for JWT invalidation (`/wazuh-qa/tests/system/test_jwt_invalidation`)
-->

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

--- a/api/test/integration/common.yaml
+++ b/api/test/integration/common.yaml
@@ -18,7 +18,6 @@ variables:
 
   # delays
   reconnect_delay: 30
-  restart_delay: 45
   restart_delay_cluster: 150
   upgrade_delay: 60
   global_db_delay: 30

--- a/api/test/integration/env/tools/print_active_agents.py
+++ b/api/test/integration/env/tools/print_active_agents.py
@@ -1,3 +1,4 @@
 from wazuh import agent
 
-print(agent.get_agents(select=['status'], filters={'status': 'active'})._affected_items)
+if __name__ == '__main__':
+    print(agent.get_agents(select=['status'], filters={'status': 'active'}).affected_items)

--- a/api/test/integration/env/tools/print_active_agents.py
+++ b/api/test/integration/env/tools/print_active_agents.py
@@ -1,0 +1,3 @@
+from wazuh import agent
+
+print(agent.get_agents(select=['status'], filters={'status': 'active'})._affected_items)

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -337,7 +337,7 @@ def test_validate_key_not_in_response(response, key):
 
 
 def check_agentd_started(response, agents_list):
-    """Wait until all the agents had their agentd process started correctly. This will avoid race conditions caused by
+    """Wait until all the agents have their agentd process started correctly. This will avoid race conditions caused by
     agents reconnections before restarting.
 
     Parameters

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -263,12 +263,14 @@ def test_validate_restart_by_node(response, data):
             affected_items.append(item['id'])
     assert response.json()['data']['affected_items'] == affected_items
     assert not response.json()['data']['failed_items']
+    healthcheck_agent_restart(response, affected_items)
 
 
 def test_validate_restart_by_node_rbac(response, permitted_agents):
     data = response.json().get('data', None)
     if data:
         if data['affected_items']:
+            healthcheck_agent_restart(response, data['affected_items'])
             for agent in data['affected_items']:
                 assert agent in permitted_agents
         else:

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -391,13 +391,12 @@ def check_agentd_started(response, agents_list):
             assert False, "The wazuh-agentd daemon was not started after requesting the restart"
 
 
-def check_agent_active_status(response, agents_list):
+def check_agent_active_status(agents_list):
     """Wait until all the agents have active status in the global.db. This will avoid race conditions caused by
     non-active agents in following test cases.
 
     Parameters
     ----------
-    response : Request response
     agents_list : list
         List of expected agents to be restarted.
     """
@@ -441,4 +440,4 @@ def healthcheck_agent_restart(response, agents_list):
     # Wait for cluster synchronization process (20 seconds)
     time.sleep(20)
     # Wait for active agent status (up to 25 seconds)
-    check_agent_active_status(response, agents_list)
+    check_agent_active_status(agents_list)

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -411,7 +411,7 @@ def check_agent_active_status(agents_list):
             command.append(python_code)
             output = subprocess.check_output(command).decode().strip()
         except subprocess.CalledProcessError:
-            assert False, f"Error while trying to get agents"
+            assert False, "Error while trying to get agents"
 
         # Transform string representation of list to list and save agents id
         id_active_agents = [agent['id'] for agent in eval(output)]

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -375,8 +375,8 @@ def check_agentd_started(response, agents_list):
                 # Save agentd logs in a list
                 command = f"docker exec env_wazuh-agent{int(agent_id)}_1 grep agentd /var/ossec/logs/ossec.log"
                 output = subprocess.check_output(command.split()).decode().strip().split('\n')
-            except subprocess.CalledProcessError:
-                assert False, f"Error while trying to get logs from agent {agent_id}"
+            except subprocess.SubprocessError as exc:
+                raise subprocess.SubprocessError(f"Error while trying to get logs from agent {agent_id}") from exc
 
             # Ignore agentd logs before restart_request_time
             logs_after_restart = [agentd_log for agentd_log in output if
@@ -389,7 +389,7 @@ def check_agentd_started(response, agents_list):
             tries += 1
             time.sleep(1)
         else:
-            assert False, "The wazuh-agentd daemon was not started after requesting the restart"
+            raise ProcessLookupError("The wazuh-agentd daemon was not started after requesting the restart")
 
 
 def check_agent_active_status(agents_list):
@@ -408,8 +408,8 @@ def check_agent_active_status(agents_list):
             # Get active agents
             output = subprocess.check_output("docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 "
                                              "/tools/print_active_agents.py".split()).decode().strip()
-        except subprocess.CalledProcessError:
-            assert False, "Error while trying to get agents"
+        except subprocess.SubprocessError as exc:
+            raise subprocess.SubprocessError("Error while trying to get agents") from exc
 
         # Transform string representation of list to list and save agents id
         id_active_agents = [agent['id'] for agent in eval(output)]
@@ -421,7 +421,7 @@ def check_agent_active_status(agents_list):
         time.sleep(1)
     else:
         non_active_agents = [a for a in agents_list if a not in id_active_agents]
-        assert False, f"Agents {non_active_agents} have a status different to active after restarting"
+        raise SystemError(f"Agents {non_active_agents} have a status different to active after restarting")
 
 
 def healthcheck_agent_restart(response, agents_list):

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -426,7 +426,7 @@ def check_agent_active_status(agents_list):
 
 
 def healthcheck_agent_restart(response, agents_list):
-    """Wait until the agent restart process of every agent in an agents list given is finished.
+    """Wait until the restart process is finished for every agent in the given list.
 
     Parameters
     ----------

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -346,6 +346,8 @@ def check_agentd_started(response, agents_list):
     agents_list : list
         List of expected agents to be restarted.
     """
+    timestamp_regex = re.compile(r'^\d\d\d\d/\d\d/\d\d\s\d\d:\d\d:\d\d')
+    agentd_started_regex = re.compile(r'agentd.*Started')
 
     def get_timestamp(log):
         """Get timestamp from log.
@@ -360,7 +362,7 @@ def check_agentd_started(response, agents_list):
         datetime
             Datetime object representing the timestamp got.
         """
-        timestamp = re.search(r'^\d\d\d\d/\d\d/\d\d\s\d\d:\d\d:\d\d', log).group(0)
+        timestamp = timestamp_regex.search(string=log).group(0)
         return datetime.strptime(timestamp, "%Y/%m/%d %H:%M:%S")
 
     # Save the time when the restart command was sent
@@ -381,8 +383,7 @@ def check_agentd_started(response, agents_list):
                                   get_timestamp(agentd_log).timestamp() >= restart_request_time.timestamp()]
 
             # Check the log indicating agentd started is in the agent's ossec.log (after the restart request)
-            if any(re.search(pattern='agentd.*Started', string=agentd_log) for agentd_log in
-                   logs_after_restart):
+            if any(agentd_started_regex.search(string=agentd_log) for agentd_log in logs_after_restart):
                 break
 
             tries += 1

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -336,9 +336,9 @@ def test_validate_key_not_in_response(response, key):
     assert all(key not in item for item in response.json()["data"]["affected_items"])
 
 
-def healthcheck_agent_restart(response, agents_list):
-    """Wait until all the agents had their agentd process terminated correctly. This will avoid race conditions caused
-    by agents reconnections before restarting.
+def check_agentd_started(response, agents_list):
+    """Wait until all the agents had their agentd process started correctly. This will avoid race conditions caused by
+    agents reconnections before restarting.
 
     Parameters
     ----------
@@ -380,12 +380,65 @@ def healthcheck_agent_restart(response, agents_list):
             logs_after_restart = [agentd_log for agentd_log in output if
                                   get_timestamp(agentd_log).timestamp() >= restart_request_time.timestamp()]
 
-            # Check the log indicating agentd was terminated is in the agent's ossec.log (after the restart request)
-            if any(re.search(pattern='agentd.*Terminated.*Received', string=agentd_log) for agentd_log in
+            # Check the log indicating agentd started is in the agent's ossec.log (after the restart request)
+            if any(re.search(pattern='agentd.*Started', string=agentd_log) for agentd_log in
                    logs_after_restart):
                 break
 
             tries += 1
             time.sleep(1)
         else:
-            assert False, "The wazuh-agentd daemon was not terminated after requesting the restart"
+            assert False, "The wazuh-agentd daemon was not started after requesting the restart"
+
+
+def check_agent_active_status(response, agents_list):
+    """Wait until all the agents have active status in the global.db. This will avoid race conditions caused by
+    non-active agents in following test cases.
+
+    Parameters
+    ----------
+    response : Request response
+    agents_list : list
+        List of expected agents to be restarted.
+    """
+    id_active_agents = []
+    tries = 0
+    while tries < 25:
+        try:
+            # Get active agents
+            python_code = "from wazuh import agent; print(agent.get_agents(select=['status'], " \
+                          "filters={'status': 'active'})._affected_items)"
+            command = "docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 -c ".split()
+            command.append(python_code)
+            output = subprocess.check_output(command).decode().strip()
+        except subprocess.CalledProcessError:
+            assert False, f"Error while trying to get agents"
+
+        # Transform string representation of list to list and save agents id
+        id_active_agents = [agent['id'] for agent in eval(output)]
+
+        if all(a in id_active_agents for a in agents_list):
+            break
+
+        tries += 1
+        time.sleep(1)
+    else:
+        non_active_agents = [a for a in agents_list if a not in id_active_agents]
+        assert False, f"Agents {non_active_agents} have a status different to active after restarting"
+
+
+def healthcheck_agent_restart(response, agents_list):
+    """Wait until the agent restart process of every agent in an agents list given is finished.
+
+    Parameters
+    ----------
+    response : Request response
+    agents_list : list
+        List of expected agents to be restarted.
+    """
+    # Wait for agentd daemon start (up to 80 seconds)
+    check_agentd_started(response, agents_list)
+    # Wait for cluster synchronization process (20 seconds)
+    time.sleep(20)
+    # Wait for active agent status (up to 25 seconds)
+    check_agent_active_status(response, agents_list)

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -347,7 +347,7 @@ def check_agentd_started(response, agents_list):
         List of expected agents to be restarted.
     """
     timestamp_regex = re.compile(r'^\d\d\d\d/\d\d/\d\d\s\d\d:\d\d:\d\d')
-    agentd_started_regex = re.compile(r'agentd.*Started')
+    agentd_started_regex = re.compile(r'agentd.+Started')
 
     def get_timestamp(log):
         """Get timestamp from log.
@@ -401,13 +401,14 @@ def check_agent_active_status(agents_list):
     agents_list : list
         List of expected agents to be restarted.
     """
+    active_agents_script_path = "/tools/print_active_agents.py"
     id_active_agents = []
     tries = 0
     while tries < 25:
         try:
             # Get active agents
-            output = subprocess.check_output("docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 "
-                                             "/tools/print_active_agents.py".split()).decode().strip()
+            output = subprocess.check_output(f"docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 "
+                                             f"{active_agents_script_path}".split()).decode().strip()
         except subprocess.SubprocessError as exc:
             raise subprocess.SubprocessError("Error while trying to get agents") from exc
 

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -406,11 +406,8 @@ def check_agent_active_status(agents_list):
     while tries < 25:
         try:
             # Get active agents
-            python_code = "from wazuh import agent; print(agent.get_agents(select=['status'], " \
-                          "filters={'status': 'active'})._affected_items)"
-            command = "docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 -c ".split()
-            command.append(python_code)
-            output = subprocess.check_output(command).decode().strip()
+            output = subprocess.check_output("docker exec env_wazuh-master_1 /var/ossec/framework/python/bin/python3 "
+                                             "/tools/print_active_agents.py".split()).decode().strip()
         except subprocess.CalledProcessError:
             assert False, "Error while trying to get agents"
 

--- a/api/test/integration/tavern_utils.py
+++ b/api/test/integration/tavern_utils.py
@@ -1,5 +1,9 @@
 import json
+import re
+import subprocess
+import time
 from base64 import b64decode
+from datetime import datetime
 from json import loads
 
 from box import Box
@@ -328,3 +332,58 @@ def test_validate_search(response, search_param):
 
 def test_validate_key_not_in_response(response, key):
     assert all(key not in item for item in response.json()["data"]["affected_items"])
+
+
+def healthcheck_agent_restart(response, agents_list):
+    """Wait until all the agents had their agentd process terminated correctly. This will avoid race conditions caused
+    by agents reconnections before restarting.
+
+    Parameters
+    ----------
+    response : Request response
+    agents_list : list
+        List of expected agents to be restarted.
+    """
+
+    def get_timestamp(log):
+        """Get timestamp from log.
+
+        Parameters
+        ----------
+        log : str
+            String representing the log to get the timestamp from.
+
+        Returns
+        -------
+        datetime
+            Datetime object representing the timestamp got.
+        """
+        timestamp = re.search(r'^\d\d\d\d/\d\d/\d\d\s\d\d:\d\d:\d\d', log).group(0)
+        return datetime.strptime(timestamp, "%Y/%m/%d %H:%M:%S")
+
+    # Save the time when the restart command was sent
+    restart_request_time = datetime.utcnow().replace(microsecond=0) - response.elapsed
+
+    for agent_id in agents_list:
+        tries = 0
+        while tries < 80:
+            try:
+                # Save agentd logs in a list
+                command = f"docker exec env_wazuh-agent{int(agent_id)}_1 grep agentd /var/ossec/logs/ossec.log"
+                output = subprocess.check_output(command.split()).decode().strip().split('\n')
+            except subprocess.CalledProcessError:
+                assert False, f"Error while trying to get logs from agent {agent_id}"
+
+            # Ignore agentd logs before restart_request_time
+            logs_after_restart = [agentd_log for agentd_log in output if
+                                  get_timestamp(agentd_log).timestamp() >= restart_request_time.timestamp()]
+
+            # Check the log indicating agentd was terminated is in the agent's ossec.log (after the restart request)
+            if any(re.search(pattern='agentd.*Terminated.*Received', string=agentd_log) for agentd_log in
+                   logs_after_restart):
+                break
+
+            tries += 1
+            time.sleep(1)
+        else:
+            assert False, "The wazuh-agentd daemon was not terminated after requesting the restart"

--- a/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
+++ b/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
@@ -143,6 +143,10 @@ stages:
         agents_list: "003,004,005"
         group_id: group1
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["003", "004"]
       status_code: 200
       json:
         error: 2
@@ -157,7 +161,6 @@ stages:
                 - "005"
           total_affected_items: 2
           total_failed_items: 1
-    delay_after: !float "{restart_delay}"
 
     # PUT /agents/group?group_id=group2
   - name: Assign all agents to group2
@@ -170,6 +173,10 @@ stages:
       params:
         group_id: group2
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "003", "004", "007"]  # 009 is not restarted as it is disconnected
       status_code: 200
       json:
         error: 0
@@ -181,7 +188,6 @@ stages:
             - "007"
             - "009"
           total_affected_items: 5
-    delay_after: !float "{restart_delay}"
 
     # PUT /agents/group?group_id=group2&force_single_group=true
   - name: Assign all agents to group2 with force_single_group
@@ -195,6 +201,11 @@ stages:
         force_single_group: True
         group_id: group2
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "002", "003", "004", "005", "006", "007", "008"]  # 009 and 010 are not restarted as
+                                                                                 # they are disconnected
       status_code: 200
       json:
         error: 0
@@ -211,7 +222,6 @@ stages:
             - "009"
             - "010"
           total_affected_items: 10
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /groups/{group_id}/configuration
@@ -230,8 +240,12 @@ stages:
       data:
         "{file_xml:s}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "002", "003", "004", "005", "006", "007", "008"]  # 009 and 010 are not restarted as
+                                                                                 # they are disconnected
       status_code: 200
-    delay_after: !float "{restart_delay}"
 
     # PUT /groups/wrong_group/configuration
   - name: Try to update configuration of a non existing group
@@ -293,7 +307,7 @@ stages:
       json:
         error: 1114
 
-  - name: Try to update configuration using a complex configuration file
+  - name: Try to update configuration using a complex configuration file (no agents in group1)
     request:
       verify: False
       url: "{protocol:s}://{host:s}:{port:d}/groups/group1/configuration"
@@ -305,7 +319,6 @@ stages:
         '{query_list_xml:s}'
     response:
       status_code: 200
-    delay_after: !float "{restart_delay}"
 
   - name: Ensure the config now has the expected value
     request:
@@ -379,7 +392,6 @@ stages:
             - "007"
             - "008"
           total_affected_items: 8
-    delay_after: !float "{restart_delay}"
 
   # PUT /groups/not_exists/restart (Group does not exist)
   - name: Restart all agents from a group which does not exist
@@ -601,7 +613,6 @@ stages:
                 - "999"
           total_affected_items: 2
           total_failed_items: 1
-    delay_after: !float "{restart_delay}"
 
     # PUT /agents/restart
   - name: Try to restart all agents
@@ -632,7 +643,6 @@ stages:
           failed_items: !anything
           total_affected_items: 8
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/{agent_id}/group/{group_id}
@@ -792,7 +802,6 @@ stages:
           affected_items:
             - "003"
           total_affected_items: 1
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/node/{node_id}/restart
@@ -831,7 +840,6 @@ stages:
         - function: tavern_utils:test_validate_restart_by_node
           extra_kwargs:
             data: "{response_data}"
-    delay_after: !float "{restart_delay}"
 
   - name: Restart all agents belonging to a node (it does not exist)
     request:

--- a/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
+++ b/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
@@ -361,6 +361,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "002", "003", "004", "005", "006", "007", "008"]
       status_code: 200
       json:
         error: 0
@@ -390,9 +394,6 @@ stages:
 
 ---
 test_name: PUT /agents/reconnect
-
-marks:
-  - xfail # Review and fix RBAC black agent API integration test - https://github.com/wazuh/wazuh/issues/10914
 
 stages:
 
@@ -582,6 +583,10 @@ stages:
       params:
         agents_list: "001,002,999"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "002"]
       status_code: 200
       json:
         error: 2
@@ -607,6 +612,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "002", "003", "004", "005", "006", "007", "008"]
       status_code: 200
       json:
         error: 0
@@ -772,6 +781,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["003"]
       status_code: 200
       json:
         error: 0

--- a/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
+++ b/api/test/integration/test_agent_PUT_endpoints.tavern.yaml
@@ -658,6 +658,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001"]
       status_code: 200
       json:
         error: 0
@@ -712,6 +716,10 @@ stages:
       params:
         force_single_group: true
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["002"]
       status_code: 200
       json:
         error: 0

--- a/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
@@ -693,10 +693,13 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
       json:
-        message: "Agent '005' removed from 'group1'."
-    delay_after: !float "{global_db_delay}"
+        message: !anystr
 
 ---
 test_name: DELETE /agents/{agent_id}/group
@@ -1244,6 +1247,10 @@ stages:
         agents_list: '001,002,005'
         group_id: group1
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
       json:
         error: 2
@@ -1295,6 +1302,10 @@ stages:
       data:
         "{file_xml:s}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
 
   - name: Try to update group3 configuration (Denied)
@@ -1647,7 +1658,6 @@ stages:
         - function: tavern_utils:test_validate_restart_by_node_rbac
           extra_kwargs:
             permitted_agents: ['003','005','007'] # Agents '003', '005' and '007' can be restarted
-    delay_after: !float "{global_db_delay}"
 
   - name: Restart all agents belonging to worker1 (Allow node_id - Could deny agent_id)
     request:
@@ -1664,7 +1674,6 @@ stages:
         - function: tavern_utils:test_validate_restart_by_node_rbac
           extra_kwargs:
             permitted_agents: ['003','005','007'] # Agents '003', '005' and '007' can be restarted
-    delay_after: !float "{global_db_delay}"
 
   - name: Restart all agents belonging to worker2 (Allow node_id - Could deny agent_id)
     request:

--- a/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
@@ -1278,7 +1278,6 @@ stages:
           failed_items: []
           total_affected_items: 1
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /groups/{group_id}/configuration
@@ -1351,7 +1350,6 @@ stages:
           total_affected_items: 1
           failed_items: []
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/reconnect
@@ -1464,7 +1462,6 @@ stages:
                 - '002'
           total_affected_items: 1
           total_failed_items: 3
-    delay_after: !float "{restart_delay}"
 
   - name: Try to restart all agents (Partially allowed, user agnostic)
     request:
@@ -1489,7 +1486,6 @@ stages:
           failed_items: []
           total_affected_items: 3
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/{agent_id}/group/{group_id}
@@ -1573,7 +1569,6 @@ stages:
           failed_items: []
           total_affected_items: 1
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/upgrade

--- a/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_black_agent_endpoints.tavern.yaml
@@ -1338,6 +1338,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
       json:
         error: 0
@@ -1351,9 +1355,6 @@ stages:
 
 ---
 test_name: PUT /agents/reconnect
-
-marks:
-  - xfail # Review and fix RBAC black agent API integration test - https://github.com/wazuh/wazuh/issues/10914
 
 stages:
 
@@ -1444,6 +1445,10 @@ stages:
       params:
         agents_list: "000,001,002,005"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
       json:
         error: 2
@@ -1469,6 +1474,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["003", "005", "007"]
       status_code: 200
       json:
         error: 0
@@ -1551,6 +1560,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["005"]
       status_code: 200
       json:
         error: 0

--- a/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
@@ -781,6 +781,10 @@ stages:
       params:
         groups_list: 'group1,group3,group2'
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["007"]
       status_code: 200
       json:
         error: 2
@@ -798,7 +802,6 @@ stages:
                 - 'group1'
           total_affected_items: 1
           total_failed_items: 2
-    delay_after: !float "{global_db_delay}"
 
 ---
 test_name: DELETE /agents/group?group_id={group_id}
@@ -899,6 +902,10 @@ stages:
         agents_list: '001,008'
         group_id: 'default'
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["001", "008"]
       status_code: 200
       json:
         error: 0
@@ -909,7 +916,6 @@ stages:
           failed_items: []
           total_affected_items: 2
           total_failed_items: 0
-    delay_after: !float "{global_db_delay}"
 
 ---
 test_name: DELETE /groups?groups_list={group_id}
@@ -1296,6 +1302,10 @@ stages:
         agents_list: '001,002,005'
         group_id: 'group2'
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["002"]
       status_code: 200
       json:
         error: 2
@@ -1310,7 +1320,6 @@ stages:
                 - '005'
           total_affected_items: 1
           total_failed_items: 2
-    delay_after: !float "{global_db_delay}"
 
   - name: Try to assign all agents to group2 (Partially allowed, user agnostic)
     request:
@@ -1322,6 +1331,10 @@ stages:
       params:
         group_id: 'group2'
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["006", "008"]
       status_code: 200
       json:
         error: 0
@@ -1332,7 +1345,6 @@ stages:
           failed_items: []
           total_affected_items: 2
           total_failed_items: 0
-    delay_after: !float "{global_db_delay}"
 
 ---
 test_name: PUT /groups/{group_id}/configuration
@@ -1350,6 +1362,10 @@ stages:
       data:
         "{file_xml:s}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["002", "006", "008"]
       status_code: 200
 
   - name: Try to update group1 configuration (Denied)
@@ -1656,6 +1672,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["008"]
       status_code: 200
       json:
         error: 0

--- a/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
@@ -1399,7 +1399,6 @@ stages:
                 - "010"
           total_failed_items: 3
         message: !anystr
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/reconnect
@@ -1570,7 +1569,6 @@ stages:
                 - '002'
           total_affected_items: 1
           total_failed_items: 3
-    delay_after: !float "{restart_delay}"
 
   - name: Try to restart all agents (Partially allowed, user agnostic)
     request:
@@ -1593,7 +1591,6 @@ stages:
           failed_items: []
           total_affected_items: 1
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/{agent_id}/group/{group_id}
@@ -1726,7 +1723,6 @@ stages:
           failed_items: []
           total_affected_items: 1
           total_failed_items: 0
-    delay_after: !float "{restart_delay}"
 
 ---
 test_name: PUT /agents/upgrade

--- a/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
+++ b/api/test/integration/test_rbac_white_agent_endpoints.tavern.yaml
@@ -1379,6 +1379,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["004"]
       status_code: 200
       json:
         error: 2
@@ -1399,9 +1403,6 @@ stages:
 
 ---
 test_name: PUT /agents/reconnect
-
-marks:
-  - xfail # Review and fix RBAC black agent API integration test - https://github.com/wazuh/wazuh/issues/10914
 
 stages:
 
@@ -1544,6 +1545,10 @@ stages:
         agents_list: "000,001,002,004"
         wait_for_complete: true  # Prevent error code 500
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["004"]
       status_code: 200
       json:
         error: 2
@@ -1575,6 +1580,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["004"]
       status_code: 200
       json:
         error: 0
@@ -1704,6 +1713,10 @@ stages:
       headers:
         Authorization: "Bearer {test_login_token}"
     response:
+      verify_response_with:
+        function: tavern_utils:healthcheck_agent_restart
+        extra_kwargs:
+          agents_list: ["004"]
       status_code: 200
       json:
         error: 0

